src/custom-element.js

```diff
@@ -443,45 +464,67 @@ function createBaseCustomElementClass(win) {
      *
      * @return {?Promise}
      * @final
+     * @restricted
      */
     buildInternal() {
       assertNotTemplate(this);
-      devAssert(this.isUpgraded(), 'Cannot build unupgraded element');
+      devAssert(this.implClass_, 'Cannot build unupgraded element');
       if (this.buildingPromise_) {
         return this.buildingPromise_;
       }
-      return (this.buildingPromise_ = new Promise((resolve, reject) => {
-        const impl = this.impl_;
+
+      this.onReadyStateInternal(ReadyState.BUILDING);
+
+      // Create the instance.
+      const implPromise = this.createImpl_();
+
+      // Wait for consent.
+      const consentPromise = implPromise.then(() => {
         const policyId = this.getConsentPolicy_();
         if (!policyId) {
-          resolve(impl.buildCallback());
-        } else {
-          Services.consentPolicyServiceForDocOrNull(this)
-            .then((policy) => {
-              if (!policy) {
-                return true;
-              }
-              return policy.whenPolicyUnblock(/** @type {string} */ (policyId));
-            })
-            .then((shouldUnblock) => {
-              if (shouldUnblock) {
-                resolve(impl.buildCallback());
-              } else {
-                reject(blockedByConsentError());
-              }
-            });
+          return;
         }
-      }).then(
+        return Services.consentPolicyServiceForDocOrNull(this)
+          .then((policy) => {
+            if (!policy) {
+              return true;
+            }
+            return policy.whenPolicyUnblock(policyId);
+          })
+          .then((shouldUnblock) => {
+            if (!shouldUnblock) {
+              throw blockedByConsentError();
+            }
+          });
+      });
+
+      // Build callback.
+      const buildPromise = consentPromise.then(() =>
+        devAssert(this.impl_).buildCallback()
+      );
+
+      // Build the element.
+      return (this.buildingPromise_ = buildPromise.then(
         () => {
-          this.preconnect(/* onLayout */ false);
           this.built_ = true;
           this.classList.add('i-amphtml-built');
           this.classList.remove('i-amphtml-notbuilt');
           this.classList.remove('amp-notbuilt');
           this.signals_.signal(CommonSignals.BUILT);
+
+          if (this.V1()) {
+            if (this.readyState_ == ReadyState.BUILDING) {
```

Comment:
It works like this: 

1. The element starts in the "building" state.
2. The implementation's `buildCallback` is called. In this callback the implementation has a chance to change the state. Most likely to "loading". It basically means: the element is built, but not yet loaded.
3. Then the control comes to this line and if the state is still "building" (the implementation hasn't changed it) then it assumes that the implementation is not interested in doing anything else and it flips the state to "complete". Otherwise it stays in the assigned state ("loading") until the implementation itself changes it.

Reply:
Thanks for the explaination! Makes sense to me